app/services/vacancy_analytics_service.rb

```diff
@@ -0,0 +1,60 @@
+class VacancyAnalyticsService
+  REDIS_KEY_PREFIX = "vacancy_referrer_stats".freeze
+
+  def self.track_visit(vacancy_id, referrer_url)
+    return if vacancy_id.blank?
+
+    # Generate a Redis key for this vacancy and referrer
+    redis_key = "#{REDIS_KEY_PREFIX}:#{vacancy_id}:#{normalize_referrer(referrer_url)}"
+    # Increment the counter in Redis
+    Redis.current.incr(redis_key)
+  end
+
+  def self.aggregate_and_save_stats
+    keys_pattern = "#{REDIS_KEY_PREFIX}:*"
+
+    Redis.current.scan_each(match: keys_pattern).each_slice(100) do |keys_batch|
+      # Creates hash that, when accessed with a missing key, assigns a new nested hash as the value. This nested hash defaults to 0 for any missing keys.
+      updates_by_vacancy = Hash.new { |h, k| h[k] = Hash.new(0) }
+      keys_to_delete = []
+
+      keys_batch.each do |key|
+        count = Redis.current.get(key).to_i
+        next if count.zero?
+
+        # Parse key to extract vacancy_id and referrer
+        _, vacancy_id, referrer = key.split(":", 3)
+
+        updates_by_vacancy[vacancy_id][referrer] += count
+        keys_to_delete << key
+      end
+
+      update_stats_in_database(updates_by_vacancy) if updates_by_vacancy.any?
+      Redis.current.del(*keys_to_delete)
+    end
+  end
+
+  def self.update_stats_in_database(vacancy_updates)
+    vacancy_updates.each do |vacancy_id, new_referrer_counts|
+      analytics = VacancyAnalytics.find_or_initialize_by(vacancy_id: vacancy_id)
+
+      merged_counts = analytics.referrer_counts.merge(new_referrer_counts) do |_, old_count, new_count|
+        old_count.to_i + new_count.to_i
+      end
+
+      analytics.referrer_counts = merged_counts
+      analytics.save!
```

Comment:
In a worst-case scenario, where somehow multiple instances of the job calling this job run at the same time, this bit could have race conditions when updating the DB data.

Some situations that come to mind and could lead to multiple `update_stats_ind_database` calls being executed at the same time: Sidekiq is busy dealing with the scheduled search alert runs, or an analytics full backfilling call, or something taking sidekiq processing off for a while...

We could avoid this by adding a lock and grouping the lock&update under a transaction:

```suggestion
    VacancyAnalytics.transaction do
      analytics = VacancyAnalytics.where(vacancy_id: vacancy_id).lock(true).first_or_initialize

      merged_counts = analytics.referrer_counts.merge(new_referrer_counts) do |_, old_count, new_count|
        old_count.to_i + new_count.to_i
      end

      analytics.referrer_counts = merged_counts
      analytics.save!
    end
```
This way, if the row is locked in for the update any other job trying to update the stats for the same vacancy will have to wait until this transaction ends, and we will ensure no data is lost!